ctapipe_io_lst/calibration.py

```diff
@@ -0,0 +1,772 @@
+                    time_shift -= time_corr[r1.selected_gain_channel, pixel_index].to_value(u.ns)
+                else:
+                    time_shift -= time_corr.to_value(u.ns)
+
+            event.calibration.tel[tel_id].dl1.time_shift = time_shift
+
+            # needed for charge scaling in ctpaipe dl1 calib
+            if selected_gain_channel is not None:
+                relative_factor = np.empty(n_pixels)
+                relative_factor[r1.selected_gain_channel == HIGH_GAIN] = self.calib_scale_high_gain.tel[tel_id]
+                relative_factor[r1.selected_gain_channel == LOW_GAIN] = self.calib_scale_low_gain.tel[tel_id]
+            else:
+                relative_factor = np.empty((n_gains, n_pixels))
+                relative_factor[HIGH_GAIN] = self.calib_scale_high_gain.tel[tel_id]
+                relative_factor[LOW_GAIN] = self.calib_scale_low_gain.tel[tel_id]
+
+            event.calibration.tel[tel_id].dl1.relative_factor = relative_factor
+
+    @staticmethod
+    def _read_calibration_file(path):
+        """
+        Read the correction from hdf5 calibration file
+        """
+        mon = MonitoringContainer()
+
+        with tables.open_file(path) as f:
+            tel_ids = [
+                int(key[4:]) for key in f.root._v_children.keys()
+                if key.startswith('tel_')
+            ]
+
+        for tel_id in tel_ids:
+            with HDF5TableReader(path) as h5_table:
+                base = f'/tel_{tel_id}'
+                # read the calibration data
+                table = base + '/calibration'
+                next(h5_table.read(table, mon.tel[tel_id].calibration))
+
+                # read pedestal data
+                table = base + '/pedestal'
+                next(h5_table.read(table, mon.tel[tel_id].pedestal))
+
+                # read flat-field data
+                table = base + '/flatfield'
+                next(h5_table.read(table, mon.tel[tel_id].flatfield))
+
+                # read the pixel_status container
+                table = base + '/pixel_status'
+                next(h5_table.read(table, mon.tel[tel_id].pixel_status))
+
+        return mon
+
+    @staticmethod
+    def load_drs4_time_calibration_file(path):
+        """
+        Function to load calibration file.
+        """
+        with tables.open_file(path, 'r') as f:
+            fan = f.root.fan[:]
+            fbn = f.root.fbn[:]
+
+        return fan, fbn
+
+    def load_drs4_time_calibration_file_for_tel(self, tel_id):
+        self.fan[tel_id], self.fbn[tel_id] = self.load_drs4_time_calibration_file(
+            self.drs4_time_calibration_path.tel[tel_id]
+        )
+
+    def get_drs4_time_correction(self, tel_id, first_capacitors, selected_gain_channel=None):
+        """
+        Return pulse time after time correction.
+        """
+
+        if self.drs4_time_calibration_path.tel[tel_id] is None:
+            return np.zeros(selected_gain_channel.shape)
```

Comment:
If selected_gain_channel is  None, this could be a problem.

Reply:
indeed, will fix.

Reply:
done